### PR TITLE
refactor: re-land #7815.

### LINKED ISSUE #7848: Re-land #7815
Re-land #7815 which was reverted in #7847 because it caused https://app.incident.io/incidents/267 .

### PR DESCRIPTION
**Pull request by crepererum**
This was some confusion with the old and new version of the pruning code. Wasn't discovered by the test because we never have delete predicates ... accept for retention (also see #7409 ).

Added tests on two layers and verified using `prod-like-querier` that the fix indeed fixes the issue and we no longer prune away everything.

Closes #7848.

### CODE REVIEW

File: iox_query_influxrpc/src/lib.rs [hunk 1]
```diff
@@ -2429,6 +2431,43 @@ mod tests {
...
+                .with_time_column()
+                .with_one_row_of_data()
+                .with_delete_predicate(Arc::new(DeletePredicate::retention_delete_predicate(1))),
+        );
+
+        let executor = Arc::new(Executor::new_testing());
+        let test_db = Arc::new(TestDatabase::new(Arc::clone(&executor)));
+        test_db.add_chunk("my_partition_key", Arc::clone(&chunk));
+
+        // verify that _field = 'foo.bar' is rewritten correctly
```
Comment (alamb) [inline iox_query_influxrpc/src/lib.rs:2452]: this comment doesn't seem quite right

File: iox_query_influxrpc/src/lib.rs [hunk 2]
```diff
@@ -2429,6 +2431,43 @@ mod tests {
...
+        let res = InfluxRpcPlanner::new(IOxSessionContext::with_testing())
+            .read_filter(Arc::clone(&test_db) as _, rpc_predicate)
+            .await
+            .expect("creating plan");
+        assert_eq!(res.plans.len(), 1);
+        let ssplan = res.plans.first().unwrap();
+        insta::assert_snapshot!(ssplan.plan.display_indent_schema().to_string(), @r###"
+        Projection: table.tag, table.field AS field, table.time [tag:Dictionary(Int32, Utf8);N, field:Float64;N, time:Timestamp(Nanosecond, None)]
+          Sort: table.tag ASC NULLS FIRST, table.time ASC NULLS FIRST [field:Float64;N, tag:Dictionary(Int32, Utf8);N, time:Timestamp(Nanosecond, None)]
+            Filter: table.tag = Dictionary(Int32, Utf8("MA")) [field:Float64;N, tag:Dictionary(Int32, Utf8);N, time:Timestamp(Nanosecond, None)]
```
Comment (alamb) [inline iox_query_influxrpc/src/lib.rs:2466]: Can you explain, perhaps in comments, why there is no time predicate (to enforce retention) in this plan?

File: influxdb_iox/tests/end_to_end_cases/querier/influxrpc/read_filter.rs [hunk 3]
```diff
@@ -981,6 +981,71 @@ async fn periods_in_predicates() {
     .await;
 }
 
+/// See https://github.com/influxdata/influxdb_iox/issues/7848
+#[tokio::test]
+async fn retention() {
```
Comment (alamb) [inline influxdb_iox/tests/end_to_end_cases/querier/influxrpc/read_filter.rs:986]: I ran this test on `main`, with the original commit (without 8abf3ae) and it fails. 🎉 

```
cd /Users/alamb/Software/influxdb_iox2 && CARGO_TARGET_DIR=/Users/alamb/Software/target-iox2 TEST_INTEGRATION=1 TEST_JDBC_INTEGRATION=1 TEST_INFLUXDB_IOX_CATALOG_DSN=postgresql://postgres@localhost:5432/postgres cargo test --test end_to_end -- retention

...

---- end_to_end_cases::querier::influxrpc::read_filter::retention stdout ----
thread 'end_to_end_cases::querier::influxrpc::read_filter::retention' panicked at 'assertion failed: `(left == right)`
  left: `["SeriesFrame, tags: _field=val,_measurement=the_table,tag=A, type: 1", "IntegerPointsFrame, timestamps: [9223372036854775807], values: \"43\""]`,
 right: `[]`: Expected:
SeriesFrame, tags: _field=val,_measurement=the_table,tag=A, type: 1
IntegerPointsFrame, timestamps: [9223372036854775807], values: "43"
Actual:
', influxdb_iox/tests/end_to_end_cases/querier/influxrpc/read_filter.rs:1033:21
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    end_to_end_cases::querier::influxrpc::read_filter::retention
```

File: iox_query_influxrpc/src/lib.rs [hunk 4]
```diff
@@ -2429,6 +2431,43 @@ mod tests {
...
+        // verify that _field = 'foo.bar' is rewritten correctly
+        let predicate = Predicate::new().with_expr("tag".as_expr().eq(lit("MA")));
+
+        let rpc_predicate = InfluxRpcPredicate::new(None, predicate);
+
+        let res = InfluxRpcPlanner::new(IOxSessionContext::with_testing())
+            .read_filter(Arc::clone(&test_db) as _, rpc_predicate)
+            .await
+            .expect("creating plan");
+        assert_eq!(res.plans.len(), 1);
```
Comment (alamb) [inline iox_query_influxrpc/src/lib.rs:2461]: This assert fails without the fix:

```
thread 'tests::test_issue_7848' panicked at 'assertion failed: `(left == right)`
  left: `0`,
 right: `1`', iox_query_influxrpc/src/lib.rs:2461:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
👍

**Review by alamb — APPROVED** [review]
LGTM -- I double checked that the tests do indeed cover the code changes and their failures are consistent with the symptoms we saw on prod.

I had some minor comments about the tests but otherwise 🚀  -- nice work

File: iox_query_influxrpc/src/lib.rs (hunk 1 above)
Comment (crepererum) [reply iox_query_influxrpc/src/lib.rs]: removed (copy+paste from another test)

File: iox_query_influxrpc/src/lib.rs (hunk 2 above)
Comment (crepererum) [reply iox_query_influxrpc/src/lib.rs:2469]: done

**Pull request merged**